src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs

```diff
@@ -1164,13 +1164,96 @@ public void Deconstruct(out int a, out int b)
+                // (7,25): error CS0306: The type 'int*' may not be used as a type argument
+                //         (int* err2, var err3) = M2();
+                Diagnostic(ErrorCode.ERR_BadTypeArgument, "err3").WithArguments("int*").WithLocation(7, 25),
+                // (8,24): error CS0306: The type 'int*' may not be used as a type argument
+                //         foreach ((int* err4, var err5) in new[] { M2() })
+                Diagnostic(ErrorCode.ERR_BadTypeArgument, "err4").WithArguments("int*").WithLocation(8, 24),
+                // (8,34): error CS0306: The type 'int*' may not be used as a type argument
+                //         foreach ((int* err4, var err5) in new[] { M2() })
+                Diagnostic(ErrorCode.ERR_BadTypeArgument, "err5").WithArguments("int*").WithLocation(8, 34),
+                // (15,17): error CS0306: The type 'int*' may not be used as a type argument
+                //         return (default(int*), default(int*));
+                Diagnostic(ErrorCode.ERR_BadTypeArgument, "default(int*)").WithArguments("int*").WithLocation(15, 17),
+                // (15,32): error CS0306: The type 'int*' may not be used as a type argument
+                //         return (default(int*), default(int*));
+                Diagnostic(ErrorCode.ERR_BadTypeArgument, "default(int*)").WithArguments("int*").WithLocation(15, 32),
+                // (25,22): error CS0306: The type 'int*' may not be used as a type argument
+                //         (int x, var (err1, y)) = (0, new C());
+                Diagnostic(ErrorCode.ERR_BadTypeArgument, "err1").WithArguments("int*").WithLocation(25, 22),
+                // (26,15): error CS0306: The type 'int*' may not be used as a type argument
+                //         (int* err2, var err3) = M2();
+                Diagnostic(ErrorCode.ERR_BadTypeArgument, "err2").WithArguments("int*").WithLocation(26, 15),
+                // (26,25): error CS0306: The type 'int*' may not be used as a type argument
+                //         (int* err2, var err3) = M2();
+                Diagnostic(ErrorCode.ERR_BadTypeArgument, "err3").WithArguments("int*").WithLocation(26, 25),
+                // (27,24): error CS0306: The type 'int*' may not be used as a type argument
+                //         foreach ((int* err4, var err5) in new[] { M2() })
+                Diagnostic(ErrorCode.ERR_BadTypeArgument, "err4").WithArguments("int*").WithLocation(27, 24),
+                // (27,34): error CS0306: The type 'int*' may not be used as a type argument
+                //         foreach ((int* err4, var err5) in new[] { M2() })
+                Diagnostic(ErrorCode.ERR_BadTypeArgument, "err5").WithArguments("int*").WithLocation(27, 34)
+                );
+        }
+
+        [Fact]
         public void NoArgIteratorTypeInAsync()
         {
             string source = @"
 using System;
 class C
 {
-    public async void M()
+    public void M()
```

Comment:
You're correct that the check is no longer specific to `async` (I saw you removed the call to `CheckRestrictedTypeInAsync` since now superseded by constraints check).

However, let's keep the test (the name could be changed to `NoArgIteratorType`). Your test above uses int pointer as "offender", whereas this test focuses on `ArgIterator`.